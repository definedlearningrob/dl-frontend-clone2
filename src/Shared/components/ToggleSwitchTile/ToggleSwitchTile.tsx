import { ChangeEvent, FC, ReactNode, SVGProps } from 'react';

import { IconContainer } from '@shared/components/IconContainer/IconContainer';
import SharedSwitch from '@shared/components/Switch/Switch';
import { cx } from '@shared/utils/cx';
import { ReactComponent as InfoIcon } from '@shared/svg/info_outlined.svg';
import { useBreakpointUp } from '@shared/hooks/useBreakpointUp';

import { Tooltip } from '../Tooltip';

type Props = {
  isEnabled: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  name: string;
  title?: string;
  additionalInfo?: string | ReactNode;
  description?: string;
  additionalSwitchLabel?: string;
  switchLabel?: string;
  alwaysEnabled?: boolean;
  Icon: FC<SVGProps<SVGSVGElement>>;
  variant?: 'sm' | 'base';
};

export const ToggleSwitchTile = ({
  isEnabled,
  name,
  title,
  additionalInfo,
  description,
  onChange,
  additionalSwitchLabel,
  switchLabel,
  alwaysEnabled,
  Icon,
  variant = 'base',
}: Props) => {
  const isFullHD = useBreakpointUp({ breakpoint: 'xxxl' });

  const isSmall = variant === 'sm';
  const isBaseSize = variant === 'base';

  const sectionClassnames = cx(
    'flex justify-between items-center gap-xs bg-white transition-colors',
    'border border-neutral-300 rounded-xs hover:border-neutral-400',
    {
      '!border-primary-500 bg-primary-200': isEnabled,
      'py-x px-sm xxxl:p-sm': isSmall,
      'p-sm xxxl:p-base': isBaseSize,
    }
  );
  const contentClassnames = cx('flex items-start flex-grow flex-shrink-0 basis-0', {
    'gap-xs xxxl:gap-sm': isSmall,
    'gap-sm': isBaseSize,
  });

  const iconClassnames = cx('text-primary-500 rounded-sm bg-primary-200 transition-colors', {
    'bg-white': isEnabled,
  });

  const titleClassnames = cx('flex items-center gap-xs mb-xxs text-xs', {
    'xxxl:text-sm': isBaseSize,
  });

  const descriptionClassnames = cx({
    'text-xxs xxxl:text-xs': isSmall,
    'text-xs xxxl:text-sm': isBaseSize,
  });

  return (
    <section className={sectionClassnames}>
      <div className={contentClassnames}>
        {Icon && (
          <IconContainer
            Icon={Icon}
            className={iconClassnames}
            paddingSize='xs'
            size={isFullHD || isBaseSize ? 'base' : 'sm'}
          />
        )}
        <div className='leading-lg'>
          {title && (
            <h6 className={titleClassnames}>
              {title}
              {additionalInfo && (
                <Tooltip message={additionalInfo}>
                  <IconContainer
                    Icon={InfoIcon}
                    className='text-primary-500'
                    paddingSize='none'
                    size='sm'
                  />
                </Tooltip>
              )}
            </h6>
          )}
          {description && <span className={descriptionClassnames}>{description}</span>}
        </div>
      </div>
      <SharedSwitch
        additionalLabel={additionalSwitchLabel}
        alwaysEnabled={alwaysEnabled}
        data-testid={`${name}-switch`}
        label={switchLabel}
        name={name}
        value={isEnabled}
        onChange={onChange}
      />
    </section>
  );
};
