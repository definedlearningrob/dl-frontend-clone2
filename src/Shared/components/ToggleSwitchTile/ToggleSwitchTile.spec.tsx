import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ToggleSwitchTile } from './ToggleSwitchTile';

describe('ToggleSwitchTile', () => {
  it('should render the section correctly', () => {
    const icon = () => <svg />;
    const { container } = render(
      <ToggleSwitchTile
        Icon={icon}
        description='Some text about the setting'
        isEnabled={true}
        name='setting'
        onChange={jest.fn}
      />
    );

    expect(screen.getByText('Some text about the setting')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should call onChange callback when toggled', () => {
    const mockedOnChange = jest.fn();
    const icon = () => <svg />;
    render(
      <ToggleSwitchTile
        Icon={icon}
        description='Some text about the setting'
        isEnabled={true}
        name='setting'
        onChange={mockedOnChange}
      />
    );

    userEvent.click(screen.getByRole('switch'));

    expect(mockedOnChange).toHaveBeenCalledTimes(1);
  });
});
