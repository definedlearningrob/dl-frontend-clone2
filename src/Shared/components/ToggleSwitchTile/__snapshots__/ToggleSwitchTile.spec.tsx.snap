// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitchTile should render the section correctly 1`] = `
<div>
  <section
    class="flex justify-between items-center gap-xs transition-colors border border-neutral-300 rounded-xs hover:border-neutral-400 !border-primary-500 bg-primary-200 p-sm xxxl:p-base"
  >
    <div
      class="flex items-start flex-grow flex-shrink-0 basis-0 gap-sm"
    >
      <div
        class="flex"
      >
        <div
          class="flex align-center justify-center grow-0 shrink-0 text-primary-500 rounded-sm transition-colors bg-white p-xs"
          data-testid="icon"
        >
          <svg />
        </div>
      </div>
      <div
        class="leading-lg"
      >
        <span
          class="text-xs xxxl:text-sm"
        >
          Some text about the setting
        </span>
      </div>
    </div>
    <label
      class="switch"
    >
      <div
        class="content"
      >
        <input
          checked=""
          class="switchInput"
          data-testid="setting-switch"
          name="setting"
          role="switch"
          type="checkbox"
        />
        <span
          class="slider"
        />
      </div>
    </label>
  </section>
</div>
`;
